lib/simple-seeder.js

```diff
@@ -0,0 +1,221 @@
+    const id = HypercoreId.encode(publicKey)
+
+    // Maybe all this belongs to a "Resource" class, including the internals _createCore/Bee/etc
+    const info = {
+      key: id,
+      type: opts.type,
+      instance: null,
+      description: opts.description || null,
+      seeders: null,
+      discovery: null,
+      blocks: { down: speedometer(), up: speedometer() },
+      network: { down: speedometer(), up: speedometer() },
+      external: opts.external !== false
+    }
+
+    if (opts.type === 'core') info.instance = await this._createCore(publicKey, info)
+    else if (opts.type === 'bee') info.instance = await this._createBee(publicKey, info)
+    else if (opts.type === 'drive') info.instance = await this._createDrive(publicKey, info, opts)
+    else if (opts.type === 'list') info.instance = await this._createBee(publicKey, info, { keyEncoding: 'utf-8', valueEncoding: 'json' })
+    else if (opts.type === 'seeders') {
+      info.type = 'drive'
+      info.instance = await this._createDrive(publicKey, info, { seeders: true })
+    } else throw new Error('Type is required')
+
+    this.resources.set(id, info)
+
+    return info
+  }
+
+  async put (key, opts) {
+    const release = await this.lock()
+    try {
+      return await this._put(key, opts)
+    } finally {
+      release()
+    }
+  }
+
+  async _put (key, opts) {
+    if (!this.has(key)) return this._add(key, opts)
+
+    const info = this.get(key)
+
+    if (info.type !== opts.type) {
+      await this._remove(key)
+      return this._add(key, opts)
+    }
+
+    if (!!info.seeders !== opts.seeders) {
+      if (opts.seeders) {
+        if (info.seeders) throw new Error('Seeders was already enabled') // Should not happen
+        const publicKey = HypercoreId.decode(info.key)
+        info.seeders = await this._createSeeders(publicKey)
+      } else {
+        await info.seeders.destroy()
+        info.seeders = null
+      }
+    }
+
+    info.description = opts.description || null
+
+    return info
+  }
+
+  async remove (key) {
+    const release = await this.lock()
+    try {
+      return await this._remove(key)
+    } finally {
+      release()
+    }
+  }
+
+  async _remove (key) {
+    const info = this.get(key)
+
+    if (info.seeders) await info.seeders.destroy()
+    if (info.discovery) await info.discovery.destroy()
+    await info.instance.close()
+
+    this.resources.delete(info.key)
+  }
+
+  has (key) {
+    const id = HypercoreId.normalize(key)
+    return this.resources.has(id)
+  }
+
+  get (key) {
+    const id = HypercoreId.normalize(key)
+    const info = this.resources.get(id)
+    if (!info) throw new Error('Resource does not exists')
+    return info
+  }
+
+  filter (onfilter) {
+    const arr = []
+    for (const info of this.values()) {
+      if (onfilter(info)) arr.push(info)
+    }
+    return arr
+  }
+
+  values () {
+    return this.resources.values()
+  }
+
+  async destroy () {
+    if (this.destroying) return
```

Comment:
Shouldn't it store the promise at the end of this method and return that instead of just returning? (or could use ready resource to avoid having to think about stuff like that)

Reply:
Sure but we don't gain anything from doing that on this case

Reply:
Ah, I had misunderstood the intent of the `if 'this.destroying` logic. I thought it was to avoid calling destroy twice, but it's to avoid `_add` being run when already destroying. If the behaviour of calling destroy twice is undefined, I'm fine with the code as-is (The current contract of calling destroy twice is that you're not sure that when the promise resolves, the instance has in fact been destroyed)

Reply:
Correction: the destroy code doesn't work, because `info` objects do not have a destroy method, so it crashes. The error gets swallowed though. I couldn't repro easily with the menu, but easy to repro by hosting a bee.

Change the goodbye definition on tracker to
```
  goodbye(async () => {
    console.log('Calling tracker destroy')
    try {
      await tracker.destroy()
    } catch (e) {
      console.error('caught error in tracker goodbye:')
      console.error(e)
    }
  })
```
Then run

`node index.js -b 7d2050aa131f1d0dad47723c3ac0a4af2dd087d4269848c9909c6b28c2463b34`

and `ctrl-c` it.

Expected output:

```^CCalling tracker destroy
destroying
caught error in tracker goodbye:
TypeError: info.destroy is not a function
    at SimpleSeeder.destroy (/home/hans/holepunch/simple-seeder/lib/simple-seeder.js:150:25)
    at Object.fn (/home/hans/holepunch/simple-seeder/index.js:62:21)
    at /home/hans/holepunch/simple-seeder/node_modules/graceful-goodbye/index.js:37:47
    at Array.map (<anonymous>)
    at loop (/home/hans/holepunch/simple-seeder/node_modules/graceful-goodbye/index.js:37:36)
    at onexit (/home/hans/holepunch/simple-seeder/node_modules/graceful-goodbye/index.js:33:3)
    at onsignal (/home/hans/holepunch/simple-seeder/node_modules/graceful-goodbye/index.js:17:3)
    at process.emit (node:events:513:28)
```

Reply:
Fwiw, it doesn't crash when using menu because menu doesn't call tracker.destroy(), but I think that's expected. Perhaps the non-menu flows also don't need to call tracker.destroy()? Then we can just get rid of the destroy logic

Reply:
Wow that's true, the `goodbye` was ignoring all errors (we should have added your PR lol, we should review that one)

If there is `info.instance` then it should do `info.instance.close || .destroy`

Reply:
i'm merging this now as i need it, feel free to fix the issue